### PR TITLE
SALTO-4861 - Salesforce: fix mapping of nested objects in deploy errors

### PR DESCRIPTION
**Pull request by tomsellek**
We need to add the types and names of nested instances for when we get deploy errors that reference them

Note that there are still error conditions in which the error mapping will be wrong. See ticket for further details.
---
~Unit tests pending.~

Before the change:
image

After the change:
image

---
_Release Notes_: 
Salesforce: Improved mapping of deploy errors to elements

---
_User Notifications_: 
N/A

### CODE REVIEW

File: packages/salesforce-adapter/src/metadata_deploy.ts [hunk 1]
```diff
@@ -385,6 +405,41 @@ const quickDeployOrDeploy = async (
...
 
+const isEmptyValue = (value: Value): boolean => {
+  if (_.isArray(value)) {
+    return value.length === 0
+  }
+  if (_.isPlainObject(value)) {
+    return Object.keys(value).length === 0
+  }
+  return value === undefined
+}
```
Comment (tomsellek) [inline packages/salesforce-adapter/src/metadata_deploy.ts:416]: There must be an existing utility function that does this, but I couldn't find it for the life of me

File: packages/salesforce-adapter/src/metadata_deploy.ts [hunk 2]
```diff
@@ -404,20 +504,13 @@ export const deployMetadata = async (
...
-    const { elemID } = getChangeData(change)
-    changeToDeployedIds[elemID.getFullName()] = deployedIds
-
-    Object.entries(deployedIds).forEach(([type, names]) => {
-      const nameToElemId: NameToElemIDMap = {}
-      names.forEach(name => {
-        nameToElemId[name] = elemID
-      })
-      // doing it in a slightly more convoluted way because deployedComponentsElemIdsByType[type] may be udefined
-      deployedComponentsElemIdsByType[type] = _.assign({}, deployedComponentsElemIdsByType[type], nameToElemId)
```
Comment (tomsellek) [inline packages/salesforce-adapter/src/metadata_deploy.ts:418]: This code was factored into `updateTypeToElemIdMapping` along with the code that handles nested types

File: packages/salesforce-adapter/src/metadata_deploy.ts [hunk 3]
```diff
@@ -81,11 +99,7 @@ const addNestedInstancesToPackageManifest = async (
   )
 
   const addNestedInstancesFromField = async (fieldName: string): Promise<MetadataIdsMap> => {
-    const rawFieldType = await (await changeElem.getType()).fields[fieldName]?.getType()
-    // We generally expect these to be lists, handling non list types just in case of a bug
-    const fieldType = isContainerType(rawFieldType)
-      ? await rawFieldType.getInnerType()
-      : rawFieldType
```
Comment (tomsellek) [inline packages/salesforce-adapter/src/metadata_deploy.ts:88]: Factored into `getTypeOfNestedElement`

File: packages/salesforce-adapter/test/adapter.crud.test.ts [hunk 4]
```diff
@@ -167,7 +167,13 @@ describe('SalesforceAdapter CRUD', () => {
             { [CORE_ANNOTATIONS.PARENT]:
               new ReferenceExpression(mockTypes.TestCustomObject__c.elemID, mockTypes.TestCustomObject__c) }
           )
-          workflowFieldUpdate = createInstanceElement(mockDefaultValues.WorkflowFieldUpdate, mockTypes.Workflow)
+          workflowFieldUpdate = createInstanceElement(
+            {
+              ...mockDefaultValues.WorkflowFieldUpdate,
+              [INSTANCE_FULL_NAME_FIELD]: 'TestCustomObject__c.TestWorkflowFieldUpdate',
```
Comment (tomsellek) [inline packages/salesforce-adapter/test/adapter.crud.test.ts:173]: This is what these instances' full names actually look like

File: packages/salesforce-adapter/test/mock_elements.ts [hunk 5]
```diff
@@ -189,7 +189,7 @@ export const mockTypes = {
   }),
   WorkflowFieldUpdate: createMetadataObjectType({
     annotations: {
-      metadataType: WORKFLOW_TASK_METADATA_TYPE,
```
Comment (tomsellek) [inline packages/salesforce-adapter/test/mock_elements.ts:192]: This was presumably a bug

File: packages/salesforce-adapter/src/metadata_deploy.ts [hunk 6]
```diff
@@ -385,6 +399,41 @@ const quickDeployOrDeploy = async (
...
+): {
+  nestedType: TypeElement
+  nestedNames: string[]
+}[] => (
+  nestedTypeInfo.nestedInstanceFields
+    .filter(field => !isEmptyValue(instance.value[field]))
+    .map(field => ({
+      nestedType: getTypeOfNestedElement(instance, field),
+      nestedNames: getNamesOfNestedElements(instance, field),
+    }))
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/metadata_deploy.ts:434]: The method `isEmptyValue` is redundant here no?

```suggestion
    .map(field => 
      nestedNames = getNamesOfNestedElements(instance, field)
      return nestedNames.length === 0
      ? undefined
      :  {
        nestedType: getTypeOfNestedElement(instance, field),
        nestedNames: getNamesOfNestedElements(instance, field),
    }).filter(isDefined)
```

**Review by tamtamirr — APPROVED** [review]
Good Work 👑  One small suggestion. Approving to save cycles.

File: packages/salesforce-adapter/src/metadata_deploy.ts (hunk 6 above)
Comment (tomsellek) [reply packages/salesforce-adapter/src/metadata_deploy.ts]: Ack, though I ended up implementing it a bit differently.

**Pull request merged**